spec/unit/util/windows/service_spec.rb

```diff
@@ -0,0 +1,667 @@
+#!/usr/bin/env ruby
+require 'spec_helper'
+
+describe "Puppet::Util::Windows::Service", :if => Puppet.features.microsoft_windows? do
+  require 'puppet/util/windows'
+
+  # the following is a helper that adds both a call to SERVICE_STATUS_PROCESS.new
+  # and QueryServiceStatusEx. What this does from the perspective of the tests is
+  # emulate would would happen from a succesful call to QueryServiceStatusEx that
+  # returned the contents of query_return. Since QueryServiceStatusEx modifies the
+  # struct that its' passed as a parameter we just set the "return" values of the
+  # struct (those that would be set after the call to QueryServiceStatusEx)
+  # when the struct is first created, and the calling process won't be the wiser
+  def expect_successful_status_query_and_return(query_return)
+    FFI::MemoryPointer.expects(:new).with(:lpword).yields(bytes_pointer)
```

Comment:
Oh wow, you mocked this stuff all out!

Usually for FFI calls we just put stuff in `integration` and makes some real calls against the OS because there's a lot to mock and there's a lot that can go wrong with bad mocks.

Reply:
Yeah, I was hoping to move as much out of integration as I could.

Reply:
Yeah, I think you might be correct that some of the tests are working through implementation rather than behavior. I am attempting to move things to unit testing as much as possible because unit testing is cheaper.

Reply:
I don't think we should hold up the PR anymore at this point, but I'm with @glennsarti on integration vs unit. The problem with all the mocks it that they're fragile, rely on internal implementation details and are typically difficult to maintain long term.

When you have logic that you want to test independently, it can often be useful to break out additional pure Ruby classes, as those are easy to unit test quickly if cleanly defined (and without implicit dependencies, which because of Rubys dynamic nature can sometimes make difficult to find).

An example of where we took this tactic recently is in the scheduled task rewrite. The original code had a single class that combined calling the COM APIs, manipulating / validating triggers (https://github.com/puppetlabs/puppet/blob/5.5.x/lib/puppet/util/windows/taskscheduler.rb#L941-L1045), and a bunch of other stuff that was very tightly coupled. It was very unwieldy to rationalize or test as a result.

In addition to a pretty light type and provider, we ended up with a couple of new classes:

* [Trigger](https://github.com/puppetlabs/puppetlabs-scheduled_task/blob/master/lib/puppet_x/puppetlabs/scheduled_task/trigger.rb) has a number of new classes that handle various types of triggers and conversions. It's easy to test in pure Ruby like https://github.com/puppetlabs/puppetlabs-scheduled_task/blob/master/spec/unit/puppet_x/puppetlabs/scheduled_task/trigger_spec.rb
* [Task](https://github.com/puppetlabs/puppetlabs-scheduled_task/blob/master/lib/puppet_x/puppetlabs/scheduled_task/task.rb) does the COM integration and handles the operations on tasks themselves (and setting some of the metadata). We were able to do a lighter integration testing layer in https://github.com/puppetlabs/puppetlabs-scheduled_task/blob/master/spec/integration/puppet_x/puppetlabs/scheduled_task/task_spec.rb because most of the code that needed to be tested was in round-tripping trigger definitions through manifests, etc. (We still have some work to do in refactoring some *other* tests)

Just some food for thought on choosing abstractions that lead to simplified testing. Usually when you have to mock a ton to validate something, it can be an opportunity to rethink these bits to avoid that.